Java release notes 3.42.0 (#206)
SAP/cloud-sdk: docs/java/release-notes.mdx

### Fixed issues
-->

## 3.42.0 - April 8, 2021

[Maven Central](https://search.maven.org/search?q=g:com.sap.cloud.sdk*%20AND%20v:3.42.0) | [Javadoc](https://help.sap.com/doc/55bb1642ea614bf582d514ed34401136/1.0/en-US/index.html)

### New Functionality

> - We improved OData V4 update requests with PATCH by allowing for adding complex properties without the need for using setters for each value inside the complex property. 

Reviewer: ```suggestion
- We improved OData v4 update requests that use the PATCH HTTP method. Now they allow adding complex properties without the need to use setters for each value inside the complex property. 
```

>   For the property you can use the builder pattern or the constructor to propagate the values.

Reviewer: ```suggestion
  You can use the builder pattern or the constructor to propagate the property values.
```

  Before:
  ```java
  Entity entity;
  ComplexProperty property = new ComplexProperty();
  property.setValue("Foo");
  entity.setComplexProperty(property);
  ```
  Now:
  ```java
  Entity entity;
  entity.setComplexProperty(new ComplexProperty("Foo"));
  ```
  Transitive changes inside a complex property will automatically be detected.

> - The OData V2 generator is able to maintain argument order in function imports and get-by-key fluent helpers over successive generations, even when the originating specification changes parameter order.

Reviewer: ```suggestion
- The OData v2 generator is able to maintain argument order in function imports and get-by-key fluent helpers over successive generations, even when the originating specification changes parameter order.
```

>   This implicit behavior can be disabled by adding the flag to the plugin configuration: `<keepExistingSignatures>false</keepExistingSignatures>`

Reviewer: ```suggestion
  This is now a default behavior. It can be disabled by adding the flag to the plugin configuration: `<keepExistingSignatures>false</keepExistingSignatures>`
```

> - We updated the OData v2 VDM for the remaining services:

Reviewer: ```suggestion
- We updated the OData v2 type-safe client (VDM) for the services that couldn't be generated because of the parameters re-ordering issue mentioned above. Now you can use them as any other service. Find the list below:
```

  - Purchase Requisition - API_PURCHASEREQ_PROCESS_SRV
  - Production Order Confirmation - API_PROD_ORDER_CONFIRMATION_2_SRV
  - Process Order Confirmation - API_PROC_ORDER_CONFIRMATION_2_SRV
  - Attachments - API_CV_ATTACHMENT_SRV
  

>   Now the SAP Cloud SDK again supports the latest state of all OData services listed in the [SAP API Business Hub for SAP S/4HANA Cloud](https://api.sap.com/package/SAPS4HANACloud?section=Artifacts).

Reviewer: ```suggestion
The collection of releases S/4HANA Cloud client libraries from the SAP Cloud SDK is now consistent with the set of OData services that you can find in the [SAP API Business Hub for SAP S/4HANA Cloud](https://api.sap.com/package/SAPS4HANACloud?section=Artifacts).
```

### Improvements

> - We enabled the OpenAPI generator to consider the vendor extension field `x-sap-cloud-sdk-api-name` in an OpenAPI specification on the operation, path or on the root level.

Reviewer: ```suggestion
- We enabled the OpenAPI generator to consider the vendor extension field `x-sap-cloud-sdk-api-name` in an OpenAPI specification on the operation, path, or on the root level.
```

> The collection of releases S/4HANA Cloud client libraries from the SAP Cloud SDK is now consistent with the set of OData services that you can find in the [SAP API Business Hub for SAP S/4HANA Cloud](https://api.sap.com/package/SAPS4HANACloud?section=Artifacts).

Reviewer: ```suggestion
The collection of released S/4HANA Cloud client libraries from the SAP Cloud SDK is now consistent with the set of OData services that you can find in the [SAP API Business Hub for SAP S/4HANA Cloud](https://api.sap.com/package/SAPS4HANACloud?section=Artifacts).
```

- We enabled the OpenAPI generator to remove the suffix `api` (regardless of the case) from the extension field value of `x-sap-cloud-sdk-api-name`, if present.
- Dependency Updates:
  - SAP dependency updates:
    - Update SCP Neo Java TomEE7 API from `1.93.10.2` to `1.94.8`
  - Other dependency updates:
    - Minor version updates:
      - Update Reactor from `3.4.3` to `3.4.4`
      - Update Protobuf from `3.15.1` to `3.15.6`
      - Update JooQ from `3.14.7` to `3.14.8`
      - Update PMD from `6.31.0` to `6.32.0`
      - Update Hibernate Core from `5.4.28.Final` to `5.4.30.Final`
      - Update Guava from `30.1-jre` to `30.1.1-jre`

>       - Update Togglz from `2.8.0` to `2.9.6`

Reviewer: ```suggestion
      - Update Togglz from `2.8.0` to `2.9.6`
      
```

  For the property you can use the builder pattern or the constructor to propagate the values.
  Before:
  ```java
  Entity entity;
  ComplexProperty property = new ComplexProperty();
  property.setValue("Foo");
  entity.setComplexProperty(property);
  ```
  Now:
  ```java
  Entity entity;
  entity.setComplexProperty(new ComplexProperty("Foo"));
  ```
  Transitive changes inside a complex property will automatically be detected.
- The OData V2 generator is able to maintain argument order in function imports and get-by-key fluent helpers over successive generations, even when the originating specification changes parameter order.
  This implicit behavior can be disabled by adding the flag to the plugin configuration: `<keepExistingSignatures>false</keepExistingSignatures>`
- We updated the OData v2 VDM for the remaining services:
  - Purchase Requisition - API_PURCHASEREQ_PROCESS_SRV
  - Production Order Confirmation - API_PROD_ORDER_CONFIRMATION_2_SRV
  - Process Order Confirmation - API_PROC_ORDER_CONFIRMATION_2_SRV
  - Attachments - API_CV_ATTACHMENT_SRV
  

>   Now the SAP Cloud SDK again supports the latest state of all OData services listed in the [SAP API Business Hub for SAP S/4HANA Cloud](https://api.sap.com/package/SAPS4HANACloud?section=Artifacts).

Author: @artemkovalyov Can you re-check your suggestion? It looks as if it would delete more content than I would expect.

### Improvements
- We enabled the OpenAPI generator to consider the vendor extension field `x-sap-cloud-sdk-api-name` in an OpenAPI specification on the operation, path or on the root level.
- We enabled the OpenAPI generator to remove the suffix `api` (regardless of the case) from the extension field value of `x-sap-cloud-sdk-api-name`, if present.
- Dependency Updates:
  - SAP dependency updates:
    - Update SCP Neo Java TomEE7 API from `1.93.10.2` to `1.94.8`
  - Other dependency updates:
    - Minor version updates:
      - Update Reactor from `3.4.3` to `3.4.4`
      - Update Protobuf from `3.15.1` to `3.15.6`
      - Update JooQ from `3.14.7` to `3.14.8`
      - Update PMD from `6.31.0` to `6.32.0`
      - Update Hibernate Core from `5.4.28.Final` to `5.4.30.Final`
      - Update Guava from `30.1-jre` to `30.1.1-jre`

>       - Update Togglz from `2.8.0` to `2.9.6`

Author: @artemkovalyov What do you mean by "We suffix matching is case insensitive."?
